### ISSUE TITLE
TLS certificate makes haproxy exit with "segfault" on startup

### ISSUE THREAD

**Issue by lukaseckert** [haproxy/haproxy#959]
## Detailed description of the problem
When the official HAProxy docker image is run with a TLS server certificate where the "X509v3 Authority Key Identifier" attributes do not contain a "keyid" attribute (but do contain 'serial' and other attributes), the haproxy process crashes with a segfault error. The same certificate does not cause any issues with other application like Node.js, RabbitMQ, etc.

## Expected behavior
Tolerant handling of an absent "keyid" attribute, or a more helpful error message instead of a crash in case HAProxy requires that attribute.

## Steps to reproduce the behavior
Pull one one the recent haproxy Docker images, apply the configuration below, mount a server certificate where the "X509v3 Authority Key Identifier" section exists, but its attributes do not specify a "keyid". In our case, there are other attributes like 'serial' or 'dirName'. Start the container, which will crash immediately.

## Do you have any idea what may have caused this?
In 
https://github.com/haproxy/haproxy/blob/1db5579bf8a31a7b5cf8d6f8790e3c3de1e59c88/src/ssl_sock.c#L6193
there is no check if akid->keyid  contains a value before passing it to ASN1_STRING_length().

## Do you have an idea how to solve the issue?
Unfortunately I do not know if "keyid" is essential for HAProxy, and if it is, why other application seem to work without that certificate attribute.

## What is your configuration?
```
global
	maxconn 100

defaults
	mode http

frontend myfrontend
	bind :443 ssl crt /etc/cert/server.pem
	bind :80
	default_backend mybackend

backend mybackend
	server s3 10.0.0.3:80
	server s4 10.0.0.3:443 ssl verify none
```

## Output of `haproxy -vv` and `uname -a`
I'm testing in the official HAProxy Docker Image. Verified with 2.2.1, 2.3.1 and 2.4-dev (always alpine)

## If HAProxy crashed: Last outputs and backtraces

```
Thread 1 (process 7):
#0  0x00007fd6ecb65eb5 in ASN1_STRING_length () from /lib/libcrypto.so.1.1
No symbol table info available.
#1  0x0000559999fb093c in ssl_get0_issuer_chain (cert=0x55999c03c140) at src/ssl_sock.c:6193
        node = <optimized out>
        hk = <optimized out>
        akid = 0x55999c04ace0
        issuer = 0x0
#2  0x0000559999fb1428 in ssl_sock_put_ckch_into_ctx (err=0x7ffea8811690, ctx=0x55999c049220, ckch=0x55999c03b6e0, path=0x55999c01ba53 "/usr/local/etc/haproxy/serverCertAndKey-merged.pem") at src/ssl_sock.c:3066
        issuer = <optimized out>
        errcode = 0
        find_chain = 0x0
        errcode = <optimized out>
        find_chain = <optimized out>
        issuer = <optimized out>
#3  ckch_inst_new_load_store (path=0x55999c01ba53 "/usr/local/etc/haproxy/serverCertAndKey-merged.pem", ckchs=ckchs@entry=0x55999c03b640, bind_conf=bind_conf@entry=0x55999c03afc0, ssl_conf=0x0, sni_filter=0x0, fcount=0, ckchi=0x7ffea880f478, err=0x7ffea8811690) at src/ssl_sock.c:3196
--Type <RET> for more, q to quit, c to continue without paging--
        ctx = 0x55999c049220
        i = <optimized out>
        order = 0
        xname = <optimized out>
        str = 0x1 <error: Cannot access memory at address 0x1>
        pkey = <optimized out>
        kinfo = <optimized out>
        names = <optimized out>
        ckch = 0x55999c03b6e0
        ckch_inst = 0x0
        errcode = 0
#4  0x0000559999fb17e0 in ssl_sock_load_ckchs (path=<optimized out>, ckchs=0x55999c03b640, bind_conf=0x55999c03afc0, ssl_conf=<optimized out>, sni_filter=<optimized out>, fcount=<optimized out>, ckch_inst=0x7ffea880f478, err=0x7ffea8811690) at src/ssl_sock.c:3325
        errcode = 0
#5  0x0000559999fb1bf0 in ssl_sock_load_cert (path=path@entry=0x55999c01ba53 "/usr/local/etc/haproxy/serverCertAndKey-merged.pem", bind_conf=bind_conf@entry=0x55999c03afc0, err=err@entry=0x7ffea8811690) at src/ssl_sock.c:3472
        buf = {st_dev = 53, st_ino = 142797, st_nlink = 1, st_mode = 33188, st_uid = 0, st_gid = 0, __pad0 = 0, st_rdev = 0, st_size = 3402, st_blksize = 4096, st_blocks = 8, st_atim = {tv_sec = 1605519447, tv_nsec = 0}, st_mtim = {tv_sec = 1605519447, tv_nsec = 0}, st_ctim = {tv_sec = 1605519650, tv_nsec = 362288400},--Type <RET> for more, q to quit, c to continue without paging--
 __unused = {0, 0, 0}}
        fp = '\000' <repeats 4096 times>
        cfgerr = 0
        ckchs = <optimized out>
        ckch_inst = 0x0
#6  0x0000559999fbc5de in bind_parse_crt (args=0x7ffea8811740, cur_arg=<optimized out>, px=<optimized out>, conf=0x55999c03afc0, err=0x7ffea8811690) at src/cfgparse-ssl.c:654
        path = '\000' <repeats 24 times>, "\374\341\022\232"...
#7  0x000055999a09b082 in cfg_parse_listen (file=0x55999c01a9e0 "/usr/local/etc/haproxy/haproxy.cfg", linenum=33, args=0x7ffea8811740, kwm=<optimized out>) at src/cfgparse-listen.c:620
        err = 0x0
        code = <optimized out>
        bind_dumped = 0
--Type <RET> for more, q to quit, c to continue without paging--
        kw = 0x55999a1ea4e8 <bind_kws+264>
        err = 0x0
        l = <optimized out>
        cur_arg = <optimized out>
        curproxy = 0x55999c0391e0
        err = <optimized out>
        rc = 0
        val = 0
        err_code = 0
        cond = 0x0
        tmplogsrv = <optimized out>
        errmsg = 0x0
        bind_conf = 0x55999c03afc0
#8  0x000055999a08bb04 in readcfgfile (file=0x55999c01a9e0 "/usr/local/etc/haproxy/haproxy.cfg") at src/cfgparse.c:2053
        status = <optimized out>
        kwm = <optimized out>
        end = <optimized out>
        line = <optimized out>
        arg = 5
        args = {0x55999c01ba40 "bind", 0x55999c01ba45 "*:443", 0x55999c01ba4b "ssl", 0x55999c01ba4f "crt", 0x55999c01ba53 "/usr/local/etc/haproxy/serverCertAndKey-merged.pem", 0x55999c01ba85 "" <repeats 60 times>}
--Type <RET> for more, q to quit, c to continue without paging--
        thisline = <optimized out>
        linesize = 2048
        f = 0x55999c01b540
        linenum = 33
        err_code = 0
        cs = <optimized out>
        pcs = 0x0
        ics = 0x55999a1ecc90 <sections>
        readbytes = 0
        outline = 0x55999c01ba40 "bind"
        outlen = 70
        outlinesize = <optimized out>
        fatal = 0
        missing_lf = -1
#9  0x000055999a0d6960 in init (argc=<optimized out>, argv=<optimized out>) at src/haproxy.c:1844
        ret = <optimized out>
        trash = 0x55999bfb5658
        arg_mode = <optimized out>
        tmp = <optimized out>
        cfg_pidfile = <optimized out>
        err_code = 0
        err_msg = 0x0
--Type <RET> for more, q to quit, c to continue without paging--
        wl = 0x55999c01a940
        progname = 0x7ffea8812e5f "haproxy"
        change_dir = <optimized out>
        px = <optimized out>
        pcf = <optimized out>
        ideal_maxconn = <optimized out>
#10 0x0000559999f9f827 in main (argc=5, argv=0x7ffea8811d78) at src/haproxy.c:3086
        err = <optimized out>
        retry = <optimized out>
        limit = {rlim_cur = 1048576, rlim_max = 1048576}
        pidfd = -1
```

## Additional information (if helpful)
(none)

**Comment by wlallemand** [comment]
Thanks for the report, that's clearly not supposed to happen, it was introduced by the auto-completion of the certificate chain when It's not provided with the PEM. I'm working on a fix.

**Comment by wlallemand** [comment]
Fixed and backported as far as 2.2.
